### ISSUE TITLE
When use MirroredStrategy for multi-gpu and fit with multi-workers there is a error --"task_done() called too many times"

### ISSUE THREAD

**Issue by smallworld-network-wupeng** [tensorflow/tensorflow#41585]
**System information**

OS: Ubuntu 18.04
Tensorflow 2.20 from pip install
Python version: 3.7.7
CUDA Version: 10.2
cuDNN Version: release 10.2, V10.2.89
GPU : 2070 x 2

**Describe the current behavior**
code:
```python
gpus = tf.config.experimental.list_physical_devices('GPU')
for g in gpus:
    tf.config.experimental.set_virtual_device_configuration(g, [tf.config.experimental.VirtualDeviceConfiguration(memory_limit=7000)])

mirrored_strategy = tf.distribute.MirroredStrategy(devices=["/gpu:0", "/gpu:1"])

train_image_generator = ImageDataGenerator(rescale=1./255, horizontal_flip=True, zoom_range=0.1, rotation_range=45) # Generator for our training data
validation_image_generator = ImageDataGenerator(rescale=1./255) # Generator for our validation data

train_data_gen = train_image_generator.flow_from_directory(batch_size=batch_size,
                                                           directory=train_dir,
                                                           shuffle=True,
                                                           target_size=(IMG_HEIGHT, IMG_WIDTH),
                                                           class_mode='binary')

val_data_gen = validation_image_generator.flow_from_directory(batch_size=batch_size,
                                                              directory=validation_dir,
                                                              target_size=(IMG_HEIGHT, IMG_WIDTH),
                                                              class_mode='binary')

sample_training_images, _ = next(train_data_gen)


with mirrored_strategy.scope():
    tinydarknet = keras.Sequential([
        keras.layers.Conv2D(16, (3, 3), strides=[1, 1], padding="same", input_shape=(IMG_HEIGHT,IMG_WIDTH, 3)),
        keras.layers.BatchNormalization(),
        keras.layers.LeakyReLU(alpha=0.1),
        keras.layers.MaxPooling2D((2, 2), strides=(2, 2)),
        ..........
        keras.layers.BatchNormalization(),
        keras.layers.AveragePooling2D(),
        keras.layers.Flatten(),
        keras.layers.Dense(1)
    ])


    # In[8]:


    tinydarknet.compile(optimizer="adam",
                 loss=tf.keras.losses.BinaryCrossentropy(from_logits=True),
                 metrics=["accuracy"])


# In[9]:

history = tinydarknet.fit(#_generator(
    train_data_gen,
    steps_per_epoch=total_train // batch_size,
    epochs=epochs,
    validation_data=val_data_gen,
    validation_steps=total_val // batch_size,
    workers=NUM_WORKERS
)

acc = history.history['accuracy']
val_acc = history.history['val_accuracy']

loss=history.history['loss']
val_loss=history.history['val_loss']

epochs_range = range(epochs)

tinydarknet.save("keras_model")
```

**Describe the expected behavior**

When I set workers=1 in fit(), it is normal work, but when I workers more than one , it is got a error:

tensorflow/core/framework/op_kernel.cc:1741] Invalid argument: ValueError: task_done() called too many times
Traceback (most recent call last):

  File "/data_ssd/anaconda3/envs/tf2/lib/python3.7/site-packages/tensorflow/python/ops/script_ops.py", line 243, in __call__
    ret = func(*args)

  File "/data_ssd/anaconda3/envs/tf2/lib/python3.7/site-packages/tensorflow/python/autograph/impl/api.py", line 309, in wrapper
    return func(*args, **kwargs)

  File "/data_ssd/anaconda3/envs/tf2/lib/python3.7/site-packages/tensorflow/python/data/ops/dataset_ops.py", line 785, in generator_py_func
    values = next(generator_state.get_iterator(iterator_id))

  File "/data_ssd/anaconda3/envs/tf2/lib/python3.7/site-packages/tensorflow/python/keras/engine/data_adapter.py", line 801, in wrapped_generator
    for data in generator_fn():

  File "/data_ssd/anaconda3/envs/tf2/lib/python3.7/site-packages/tensorflow/python/keras/utils/data_utils.py", line 880, in get
    six.reraise(*sys.exc_info())

  File "/data_ssd/anaconda3/envs/tf2/lib/python3.7/site-packages/six.py", line 703, in reraise
    raise value

  File "/data_ssd/anaconda3/envs/tf2/lib/python3.7/site-packages/tensorflow/python/keras/utils/data_utils.py", line 875, in get
    self.queue.task_done()

  File "/data_ssd/anaconda3/envs/tf2/lib/python3.7/queue.py", line 74, in task_done
    raise ValueError('task_done() called too many times')

ValueError: task_done() called too many times

**Comment by Saduf2019** [comment]
@smallworld-network-wupeng 
I ran the code shared and its unindented, please share simple stand alone indented code such that we can replicate the error faced or if possible please share a colab gist with the error.

**Comment by smallworld-network-wupeng** [comment]
> @smallworld-network-wupeng
> I ran the code shared and its unindented, please share simple stand alone indented code such that we can replicate the error faced or if possible please share a colab gist with the error.

fixed

**Comment by Saduf2019** [comment]
@smallworld-network-wupeng 
Kindly provide all [details to replicate the issue](https://colab.research.google.com/gist/Saduf2019/00626c9969693c68618cec46623dd509/untitled291.ipynb) or share a colab gist with the error faced.

**Comment by smallworld-network-wupeng** [comment]
> @smallworld-network-wupeng
> Kindly provide all [details to replicate the issue](https://colab.research.google.com/gist/Saduf2019/00626c9969693c68618cec46623dd509/untitled291.ipynb) or share a colab gist with the error faced.

I put all code to this colab. But how to upload the train images ?

**Comment by Saduf2019** [comment]
@smallworld-network-wupeng 
I do not have access to the link shared, please attach the images in a  zipped folder.

**Comment by quassy** [comment]
I have the same error with my code, I can't share it but could try to replicate it with the code and images of @smallworld-network-wupeng once fully provided.

```
OS: Ubuntu 20.04
Tensorflow 2.2.0 from pip install
Python version: 3.8.2
CUDA Version: 10.1, V10.1.243 (nvcc --version)
GPU: Titan RTX x 2
```

**Comment by smallworld-network-wupeng** [comment]
> I have the same error with my code, I can't share it but could try to replicate it with the code and images of @smallworld-network-wupeng once fully provided.
> 
> ```
> OS: Ubuntu 20.04
> Tensorflow 2.2.0 from pip install
> Python version: 3.8.2
> CUDA Version: 10.1, V10.1.243 (nvcc --version)
> GPU: Titan RTX x 2
> ```
> 
> #39933 might fix the issue, is it included in the current release candidates for 2.3.0?

Yes , it is same bug , I fixed it.

**Comment by nikitamaia** [comment]
Hi @smallworld-network-wupeng, glad to hear you were able to fix the issue and get your model training. 
This slowdown can happen if your program is input bound and your GPUs aren't being used as efficiently as possible (there are other possibilities too, but this one is common). I would suggest taking a look at this tutorial for the [Tensorboard Profiler](https://www.tensorflow.org/tensorboard/tensorboard_profiling_keras) which can help to discover what inefficiencies you might have. If you can share fully reproducible code, I'd be happy to help diagnose further.

Also since @quassy had the same original issue, do you might explaining how you fixed it? Just in case someone else runs into the same problem in the future.

**Comment by quassy** [comment]
@smallworld-network-wupeng Could you share what you have changed?

I personally tested my code with tensorflow==2.3.0rc2 just now and it seems to run without this error now. Will report back if the error comes up again.

**Comment by smallworld-network-wupeng** [comment]
> @smallworld-network-wupeng Could you share what you have changed?
> 
> I personally tested my code with tensorflow==2.3.0rc2 just now and it seems to run without this error now. Will report back if the error comes up again.

yes, it is looks fixed in 2.3.0rc2 . You can fixed it as [#39933 ](https://github.com/tensorflow/tensorflow/pull/39933) if you use 2.2.0
